qforce/forcefield.py

```diff
@@ -387,89 +394,198 @@ def set_charge(self, non_bonded):
+
+
+    def write_frcmod_dihedrals(self, frcmod, terms, j2cal):
+        if len(terms['dihedral']) > 0:
+            frcmod.write("\nDIHE\n")
+
+        # rigid dihedrals
+        if len(terms['dihedral/rigid']) > 0:
+
+            for dihed in terms['dihedral/rigid']:
+                ids = dihed.atomids + 1
+                equ = dihed.equ
+
+                if dihed.equ == 0.00 or dihed.equ == 3.141592653589793:
+                    equ = 180.0
+
+                fconst = dihed.fconst * j2cal
+
+                fconstAmb = (2*fconst/(2**2))
+                frcmod.write(f"{ids[0]:<2}-{ids[1]:<2}-{ids[2]:<2}-{ids[3]:<2}")
+                frcmod.write(f" 2 {fconstAmb:>15.2f} {equ:>15.2f} 2\n")
+
+        if len(terms['dihedral/flexible']) > 0:
+
+            # flexible dihedrals
+            for dihed in terms['dihedral/flexible']:
+                ids = dihed.atomids + 1
+                c = dihed.equ * j2cal
+                fc = [1.0*c[0] + 0.5*c[2] + 0.3750*c[4],
+                      1.0*c[1] + 0.75*c[3] + 0.6250*c[5],
+                      0.5*c[2] + 0.5*c[4],
+                      0.25*c[3] + 0.3125*c[5],
+                      0.125*c[4],
+                      0.0625*c[5]]
+
+                for n in range(3, 0, -1):
+                    if n % 2 == 1:
+                        equ = 180.0
+                    else:
+                        equ = 0.00
+                    frcmod.write(f"{ids[0]:<2}-{ids[1]:<2}-{ids[2]:<2}-{ids[3]:<2}")
+                    frcmod.write(f" 2 {fc[n]:>15.2f} {equ:>15.2f} -{n+1}\n")
+
+                # Last term
+                frcmod.write(f"{ids[0]:<2}-{ids[1]:<2}-{ids[2]:<2}-{ids[3]:<2}")
+                frcmod.write(f" 2 {fc[0]:>15.2f} {0.00:>15.2f}  1\n")
+
+        # inversion dihedrals
+        if len(terms['dihedral/inversion']) > 0:
+
+            for dihed in terms['dihedral/inversion']:
+                ids = dihed.atomids + 1
+                c0, c1, c2 = convert_to_inversion_rb(dihed.fconst, dihed.equ)
+
+                fc = [(1.0*c0/2 + 0.5*c2/2)* j2cal,
+                      1.0*c1/2* j2cal,
+                      0.5*c2/2* j2cal]
+
+                for n in range(3, 0, -1):
+                    if n % 2 == 1:
+                        equ = 180.0
+                    else:
+                        equ = 0.00
+                    frcmod.write(f"{ids[0]:<2}-{ids[1]:<2}-{ids[2]:<2}-{ids[3]:<2}")
+                    frcmod.write(f" 1 {fc[n]:>15.2f} {equ:>15.2f} -{n+1}\n")
+
+                # Last term
+                frcmod.write(f"{ids[0]:<2}-{ids[1]:<2}-{ids[2]:<2}-{ids[3]:<2}")
+                frcmod.write(f" 1 {fc[0]:>15.2f} {0.00:>15.2f}  1\n")
+
+
+            # improper dihedrals
+            if len(terms['dihedral/improper']) > 0:
+                frcmod.write("\nIMPROPER\n")
+                for dihed in terms['dihedral/improper']:
+                    ids = dihed.atomids + 1
+                    equ = np.degrees(dihed.equ)
```

Comment:
Amber only accept 180 as the equilibrium value for improper
`Warning: Expected Improper Torsion PHASE=180 (0.000000)`